Fix TrayIcon menu crash

logiclrd opened the pull request:
This PR addresses a crash that occurs when activating TrayIcon menu items. It contains two commits:

* 4818fa9: Adds a new `samples` project called `TrayIcon` which _should_ work. It follows all the guidelines, but with the current Avalonia code it crashes as soon as any tray icon menu item is activated.
* bb3f939: Fixes the crash observed in the `TrayIcon` sample by thunking to the UI thread when processing a `NativeMenuItem.CanExecuteChanged` notification.

## What is the current behavior?
At least on Ubuntu 24.04 running Xorg, activating a tray icon menu item causes the process to crash immediately with the exception:
```
Unhandled exception. System.InvalidOperationException: Call from invalid thread
```
This is described in detail in #15940.

## What is the updated/expected behavior with this PR?
With the fix to `NativeMenuItem.cs`, tray icon menu items can now be activated without the process crashing.

## How was the solution implemented (if it's not obvious)?
`Dispatcher.UIThread.Invoke`

## Checklist

- [n/p] Added unit tests (if possible)?
- [n/a] Added XML documentation to any related classes?
- [n/a] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Fixed issues
Fixes #15940

logiclrd committed:
Updated CanExecuteChanged in NativeMenuItem.cs to thunk to the UI thread if it ends up running on a non-UI thread.

cla-avalonia commented:
@logiclrd,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

logiclrd commented:
@cla-avalonia agree

logiclrd commented:
(rebase)

maxkatz6 commented on samples/TrayIcon/AboutWindow.axaml:
```diff
@@ -0,0 +1,18 @@
+<Window xmlns="https://github.com/avaloniaui"
```
I am not sure about adding yet another project here.
Ideally, TrayIcon test should be added to the IntegrationTestApp project + new e2e tests written per each platform (can be done separately).

logiclrd replied:
Hmm, okay. Coming at this from the other direction, when I first started writing my application, I found all existing documentation of `TrayIcon` to be incomplete and there were no samples. This would be _the first_ sample showing how to use `TrayIcon` as far as I am aware. It serves to illustrate the crash in this PR, but after the PR is merged, it serves as a `TrayIcon` example for new developers.

logiclrd replied:
Is there a coded UI test infrastructure in place that would allow automated testing of this? If so then it's also a good point to add a test of this crash to that as well. :-)

logiclrd replied:
Hmm, I duplicated the `TrayIcon` stuff into `IntegrationTestApp` but no menu appears when I click on the tray icon. I tried a bunch of different ways to make `IntegrationTestApp` be more like `TrayIcon` but I wasn't able to figure out what the difference is.

logiclrd replied:
I see project `Avalonia.IntegrationTests.Appium`. Does Appium support interacting with tray icons? I tried some searches related to this and couldn't find anything conclusive. I've never worked with Appium before.

logiclrd replied:
I am not familiar with Appium coded UI tests, and to me it feels like a significant increase in scope to replace a straightforward sample app with automated tests. Perhaps that could be captured in a backlog for someone with the expertise and handled independently from this PR?

logiclrd replied:
> I am not sure about adding yet another project here.

Regarding this point, I personally feel like it makes sense, at least from my outsiders' perspective. The folder is called `samples`. I am a new developer. I want to know how to do a tray icon. I stumble across `samples`. If one of them says `TrayIcon`, then I know exactly where to go. If not, then how am I supposed to just magically know that `IntegrationTestApp` is what I should be looking at? And how is an integration test test app a good sample of how the functionality works??

maxkatz6 replied:
> This would be the first sample 

ControlCatalog has one. But yeah, we should have proper docs and Avalonia.Samples as well.

`IntegrationTestApp` is a normal Avalonia app, so tray icon should work there. `Avalonia.IntegrationTests.Appium` is a e2e testing project which runs `IntegrationTestApp` app, and automatically manipulates with the app to check all the test cases. 

> Does Appium support interacting with tray icons

That's a good question. It likely does, but might be tricky to get correct automation right.

maxkatz6 replied:
> Perhaps that could be captured in a backlog for someone with the expertise and handled independently from this PR?

Yeah, as I mentioned in the first message, e2e can be done separately. But let's avoid another project in this PR too.

logiclrd replied:
Hmm, okay. That project is serving a purpose presently, though -- it's fulfilling the "first demonstrate the problem in one commit" part before the next commit fixes the problem. You want I should edit that commit out anyway, so that this PR is just a bug fix out of the blue with no proof?

maxkatz6 replied:
If you would prefer to have any sample, please add it to the `IntegrationTestApp` project. Without any tests, just a tray icon. Somebody else, probably me, will add tests on top of that later.

logiclrd replied:
Hmm, well, I tried to merge it in there, but when I run the `IntegrationTestApp` project, the menu doesn't show up when clicking on the tray icon. Not sure how it's different from the `TrayIcon` sample, but there must be something...
```
    <TrayIcon.Icons>
        <TrayIcons>
            <TrayIcon Icon="/Assets/icon.ico">
                <TrayIcon.Menu>
                    <NativeMenu>
                        <NativeMenuItem Header="Show _Test Window" Command="{Binding ShowWindowCommand}" />
                    </NativeMenu>
                </TrayIcon.Menu>
            </TrayIcon>
        </TrayIcons>
    </TrayIcon.Icons>
```

maxkatz6 replied:
Works here. Right click. But I removed command as it doesn't have view model locally.

![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/6b6dc308-9712-4622-ad62-46a6788d0ab9)

logiclrd replied:
That's very strange. It continues to not work for me. I tried removing the command binding just in case that was why but it didn't fix it. Since I can't get the menu to show up, I can't confirm the crash with `IntegrationTestApp`. It's a technicality but I would feel better if you could confirm the crash without c71d999 and the fix with it.

logiclrd replied:
Or, rather, with and without 04e3225ff. :-)

logiclrd replied:
Okay I switched from GNOME Desktop to KDE Plasma and now the menu shows up, and I can confirm the fix. :-) Definitely crashes without 04e3225, definitely fixed with, at least on my end.

logiclrd committed:
Updated the TrayIcon in IntegrationTestApp to have an associated menu.
Updated App.axaml.cs to set the DataContext for binding and provide a command implementation.
Added a project reference to Avalonia.ReactiveUI.

logiclrd commented:
(Restored the command binding, because it does in fact work. :-) )

maxkatz6 committed:
Replace ReactiveUI with MiniMvvm to simplify integration tests app
